src/validation.cpp

```diff
@@ -417,21 +418,135 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationSt
- */
-static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool& pool, CValidationState& state, const CTransactionRef& ptx,
-                              bool* pfMissingInputs, int64_t nAcceptTime, std::list<CTransactionRef>* plTxnReplaced,
-                              bool bypass_limits, const CAmount& nAbsurdFee, std::vector<COutPoint>& coins_to_uncache, bool test_accept) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+class MemPoolAccept
+{
+public:
+    MemPoolAccept(CTxMemPool& mempool) : pool(mempool), view(&dummy), viewMemPool(pcoinsTip.get(), pool) {
+        nLimitAncestors = gArgs.GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
+        nLimitAncestorSize = gArgs.GetArg("-limitancestorsize", DEFAULT_ANCESTOR_SIZE_LIMIT)*1000;
+        nLimitDescendants = gArgs.GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT);
+        nLimitDescendantSize = gArgs.GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000;
+    }
+
+    /**
+     * Single transaction acceptance
+     *
+     * @param[out] coins_to_uncache   Return any outpoints which were not previously present in the
+     *                                coins cache, but were added as a result of validating the tx
+     *                                for mempool acceptance. This allows the caller to optionally
+     *                                remove the cache additions if the associated transaction ends
+     *                                up being rejected by the mempool.
+     */
+    bool AcceptSingleTransaction(const CChainParams& chainparams, CTxMemPool& pool,
+            CValidationState& state, const CTransactionRef& ptx,
+            bool* pfMissingInputs, int64_t nAcceptTime,
+            std::list<CTransactionRef>* plTxnReplaced, bool bypass_limits,
+            const CAmount& nAbsurdFee,
+            std::vector<COutPoint>& coins_to_uncache, bool test_accept);
+
+private:
+    // We put the arguments we're handed into a struct, so we can pass them
+    // around easier.
+    struct ATMPArgs {
+        const CChainParams& chainparams;
+        CValidationState &state;
+        bool* pfMissingInputs;
+        int64_t nAcceptTime;
+        std::list<CTransactionRef>* plTxnReplaced;
+        bool bypass_limits;
+        const CAmount& nAbsurdFee;
+        std::vector<COutPoint>& coins_to_uncache;
+        bool test_accept;
+    };
+
+    // All the intermediate state that gets passed between the various levels
+    // of checking a given transaction.
+    struct Workspace {
+        Workspace(const CTransactionRef& ptx) : hash(ptx->GetHash()) {}
+        std::set<uint256> setConflicts;
+        CTxMemPool::setEntries allConflicting;
+        CTxMemPool::setEntries setAncestors;
+        std::unique_ptr<CTxMemPoolEntry> entry;
+
+        bool fReplacementTransaction{false};
+        CAmount nModifiedFees{0};
+        CAmount nConflictingFees{0};
+        size_t nConflictingSize{0};
```

Comment:
I'd prefer not to initialize the memory, so that valgrind and other memory tools can warn about logic errors in the code.